src/backend/cdb/cdbmutate.c

```diff
@@ -1014,6 +1014,17 @@ shareinput_mutator_xslice_2(Node *node, PlannerInfo *root, bool fPop)
 					break;
 			}
 		}
+
+		/*
+		 * Remove writer gang for the shareinput scan readers in case
+		 * of modifying CTE
+		 */
+		if (root->parse->hasModifyingCTE &&
+			currentSlice->gangType == GANGTYPE_PRIMARY_WRITER
+			&& !plan->lefttree)
+		{
+			currentSlice->gangType = GANGTYPE_PRIMARY_READER;
+		}
```

Comment:
One more error
```
explain (slicetable,costs off) with cte as (insert into t_hashed select i, i*2 from generate_series(1,5) i  returning *) select * from t2 join (select * from cte union all select * from cte) x  using (i);
                                              QUERY PLAN                                              
------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)
   ->  Hash Join
         Hash Cond: (t2.i = "*SELECT* 1".i)
         ->  Seq Scan on t2
         ->  Hash
               ->  Redistribute Motion 3:3  (slice2; segments: 3)
                     Hash Key: "*SELECT* 1".i
                     ->  Append
                           ->  Subquery Scan on "*SELECT* 1"
                                 ->  Subquery Scan on cte
                                       ->  Shared Scan (share slice:id 2:0)
                                             ->  Insert on t_hashed
                                                   ->  Redistribute Motion 1:3  (slice3; segments: 1)
                                                         Hash Key: i.i
                                                         ->  Function Scan on generate_series i
                           ->  Subquery Scan on "*SELECT* 2"
                                 ->  Subquery Scan on cte_1
                                       ->  Shared Scan (share slice:id 2:0)
 Optimizer: Postgres-based planner
 Slice 0: Dispatcher; root 0; parent -1; gang size 0
 Slice 1: Reader; root 0; parent 0; gang size 3
 Slice 2: Primary Writer; root 0; parent 1; gang size 3
 Slice 3: Singleton Reader; root 0; parent 2; gang size 1; segment 1
(23 rows)
```

With patch
```
explain (slicetable,costs off) with cte as (insert into t_hashed select i, i*2 from generate_series(1,5) i  returning *) select * from t2 join (select * from cte union all  select * from cte) x  using (i);
                                              QUERY PLAN                                              
------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)
   ->  Hash Join
         Hash Cond: (t2.i = "*SELECT* 1".i)
         ->  Seq Scan on t2
         ->  Hash
               ->  Redistribute Motion 3:3  (slice2; segments: 3)
                     Hash Key: "*SELECT* 1".i
                     ->  Append
                           ->  Subquery Scan on "*SELECT* 1"
                                 ->  Subquery Scan on cte
                                       ->  Shared Scan (share slice:id 2:0)
                                             ->  Insert on t_hashed
                                                   ->  Redistribute Motion 1:3  (slice3; segments: 1)
                                                         Hash Key: i.i
                                                         ->  Function Scan on generate_series i
                           ->  Subquery Scan on "*SELECT* 2"
                                 ->  Subquery Scan on cte_1
                                       ->  Shared Scan (share slice:id 2:0)
 Optimizer: Postgres-based planner
 Slice 0: Dispatcher; root 0; parent -1; gang size 0
 Slice 1: Reader; root 0; parent 0; gang size 3
 Slice 2: Reader; root 0; parent 1; gang size 3
 Slice 3: Singleton Reader; root 0; parent 2; gang size 1; segment 1
(23 rows)
```

Reply:
Figured out that I erroneously supposed that `shareinput_mutator_xslice_2()` works only at cross slices